Improve PPP Authentication support

jsebechlebsky opened the pull request:
Hello,
this pull request improves support for PPP authentication methods:
- binds EAP layer to PPP
- Adds support for PAP (RFC1334) + tests
- Adds support for CHAP (RFC1994) + tests

gpotter2 commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   FieldLenField("value_size", None, fmt="B", length_of="value"),
+                   XStrLenField("value", "\0"*8, length_from=lambda p: p.value_size),
```
Hey, could you set `b"\0"` here instead of `"\0"` ?

jsebechlebsky replied:
Of course, I've changed that..

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+                return PPP_PAP_Request
+            elif o in [2,3]:
+                return PPP_PAP_Response
+        return cls
```
You might want to handle the case where a `code` has been provided in `kargs`. Maybe you could have a look at `LLTDAttribute.dispatch_hook()` in `scapy/layers/lltd.py`?

jsebechlebsky replied:
Thanks for your review. I've checked the code you've referred to - that's such a cool trick! I'll implement that and modify pull request.

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+                   StrLenField("password", None, length_from=lambda p: p.passwd_len)]
+
+    def mysummary(self):
+        return "PAP-Request username='{0}' password='{1}'".format(self.username, self.password)
```
Could you use `self.sprintf()`?

jsebechlebsky replied:
No problem, I'll change that.

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+    def mysummary(self):
+        res = "PAP-Ack" if self.code == 2 else "PAP-Nak"
+        if self.msg_len > 0:
+            res += " msg='{0}'".format(self.message)
```
Same remark here (and in the next `.mysummary()` implementations) about using `sprintf()`.

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+               and self.code in (3, 4) and self.id == other.id
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
```
Same remark here about `.dispatch_hook()` implementation.

jsebechlebsky commented:
Hi, I've modified the pull request, could you review the changes and merge it if everything is OK?

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+    name = "PPP Password Authentication Protocol"
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id",0),
+                   ShortField("len", None),
```
This should probably be a `FieldLenField("len", None, length_of="data", fmt="!H")`. Then you don't need your specific `post_build()`.

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+class PPP_PAP_Request(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
```
This should be:
```Python
    FieldLenField("len", None, length_of="username", fmt="!H",
                  adjust=lambda pkt, val: val + 2 + len(pkt.password))
```

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+class PPP_PAP_Response(PPP_PAP):
+    fields_desc = [ByteEnumField("code", 2, _PPP_paptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
```
This should be:
```Python
    FieldLenField("len", None, length_of="username", fmt="!H",
                  adjust=lambda _, val: val + 1)
```

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+    fields_desc = [ByteEnumField("code", 1, _PPP_chaptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
+                   ConditionalField(StrLenField("data","",length_from=lambda p: p.len-4),
```
Since both fields are `StrLenField`, I think it would be better to remove these `ConditionalField` and use the same name in both cases. What do you think?
Also, this would ease the use of a `FieldLenField` for `"len"`.

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+
+    def mysummary(self):
+        if self.code == 3:
+            return "CHAP Success message='{0}'".format(self.message)
```
Could you use `self.sprintf()`?

p-l- commented on scapy/layers/ppp.py:
```diff
@@ -552,18 +552,163 @@ class PPP_LCP_Discard_Request(PPP_LCP):
+class PPP_CHAP_ChallengeResponse(PPP_CHAP):
+    fields_desc = [ByteEnumField("code", 1, _PPP_chaptypes),
+                   XByteField("id", 0),
+                   ShortField("len", None),
```
There again, I think it would probably be better to use a `ShortField`.

p-l- commented:
@jsebechlebsky ping :)

jsebechlebsky commented:
Sorry for such a delay, I'll fix the issues soon ;)

jsebechlebsky commented:
I'm again sorry it took so long. I've made the changes you've requested. The last commit fixes import of EAP, which was moved to separate eap.py file meanwhile -> I wonder how it passed the tests in pptp.uts, the change should have broke them.

p-l- commented:
I think you no longer use some imports (e.g., `ShortField` at least). Could you check with for example pylint? Also, could you try to comply to PEP-8 (add a space after `:` and `,`, e.g. `lambda _,x:x+4` => `lambda _, x: x + 4`? Thanks!

jsebechlebsky commented:
`ShortField` is still used in the affected file `ppp.py`. I've added commit removing only unused import which is `LenField`. Added code should be PEP-8 compliant now, however some lines are longer than 79 characters, I left it that way,since it seems quite common in the codebase. However, if you want me to change it, I can.

p-l- commented:
Thanks! Great job!
